File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.java
// 그냥 캐시 로직만 구현하면 되는 문제
// 입력이 10만이라 선형시간으로 끝내기
// 캐시 되었을 경우는 +1, 아닐 경우는 +5
// 캐시 리스트는 한칸씩 밀어내야 하기 때문에 Deque로 간단하게 구성
// cacheSize 이상이면 삭제하는 로직 추가 or 사이즈 지정
// 그냥 Deque 쓸 줄 아는지 보는 문제, 일반 배열로도 충분히 구현 가능
// 주의사항 1. 캐시된 도시이름 조회시 맨 앞으로 빼줄 것
// 주의사항 2. 대소문자 구분을 하지 않는 점 주의
import java.util.*;
class Solution {
    public int solution(int cacheSize, String[] cities) {
        if(cacheSize == 0) return cities.length*5;
        int runtime = 0;
        Deque<String> cache = new ArrayDeque<>();
        for(String el : cities){
            String city = el.toLowerCase();
            if(cache.contains(city)){
                cache.remove(city);
                runtime++;
            }else{
                if(cache.size() >= cacheSize){
                    cache.removeLast();
                }
                runtime+=5;
            }
            cache.addFirst(city );
        }
        return runtime;
    }
}

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.js
function solution(cacheSize, cities) {
    // 첫 번째 풀이 시간 복잡도 O(n)
    if(cacheSize === 0) return cities.length*5;
    
    cities = cities.map(el => el.toLowerCase());
    let stack = Array(cacheSize).fill('');
    let time = 0;
    for(city of cities){
        if(stack.includes(city)){
            let test = stack.filter(el => el !== city)
            stack = [...test,city];
            time++;
        }else{
            stack = [...stack.slice(1),city];
            time+=5;
        }
    }
    
    return time;
//      // 다른 분의 풀이를 보고 리팩토링
//     const MISS = 5, HIT = 1;

//     if (cacheSize === 0) return MISS * cities.length;

//     let answer = 0,
//         cache = [];

//     cities.forEach(city => {
//         city = city.toLowerCase();

//         let idx = cache.indexOf(city);

//         if (idx > -1) {
//             cache.splice(idx, 1);
//             answer += HIT;
//         } else {
//             if (cache.length >= cacheSize) cache.shift();
//             answer += MISS;
//         }

//         cache.push(city);
//     });

//     return answer;
}

// 테스트 케이스2 예시
// [,,제주]5
// [,제주,판교]5
// [제주,판교,서울]5
// [판교,서울,제주]1
// [서울,제주,판교]1
// [제주,판교,서울]1
// [판교,서울,제주]1
// [서울,제주,판교]1
// [제주,판교,서울]1